src/Sulu/Component/Content/Types/BlockContentType.php

```diff
@@ -392,6 +400,42 @@ private function prepareData(PropertyInterface $property, callable $dataCallback
                 }
             }
 
+            if ($returnType
+                && \is_array($blockPropertyTypeSettings)
+                && isset($blockPropertyTypeSettings['schedules_enabled'])
+                && $blockPropertyTypeSettings['schedules_enabled']
+                && isset($blockPropertyTypeSettings['schedules'])
+                && \is_array($blockPropertyTypeSettings['schedules'])
+            ) {
+
+                $now = \DateTime::createFromFormat('U', time());
+                $soonestFutureSchedule = null;
+
+                foreach ($blockPropertyTypeSettings['schedules'] as $schedule) {
+                    switch ($schedule['type']) {
+                        case 'fixed':
+                            $start = new \DateTime($schedule['start']);
+                            if ($start > $now && (!$soonestFutureSchedule || $start < $soonestFutureSchedule)) {
+                                $soonestFutureSchedule = $start;
+                            }
+                            $end = new \DateTime($schedule['end']);
+                            if ($end > $now && (!$soonestFutureSchedule || $end < $soonestFutureSchedule)) {
+                                $soonestFutureSchedule = $end;
+                            }
+                            break;
+                        case 'weekly':
+                            // TODO
+                            break;
+                    }
+                }
```

Comment:
Should we move this into an own service for have a better way to create Unit Tests for this and reuse the logic?

Reply:
Was wondering about that as well... That's also one of the reasons I was skeptical about the `BlockSkipperInterface`. We could integrate that in the `ScheduleBlockSkipperInterface`, but that seems weird, because the name doesn't suggest that anything like that is happening. The other opportunity would be to create a new service (with or without interface), but it would be nice to have one interface that could also do stuff like that. Maybe we are talking about a `BlockIterator` here? :thinking:

Reply:
When thinking about it, just this part we are talking about here, would be a nice use case for the `Visitor` pattern... We are just doing something based on that information, but do not want to add that to the data structure we are iterating over.

But I still think that it would be nice to combine that in a way, that we can skip a block (that's why I came up with a `BlockIterator`, because in Java you can call `remove` on an `Iterator`, but it does not really match here, because we want to extract some logic, for which the `BlockIterator` is not the right tool.

If we don't find a nice way to combine these behaviors, I would probably stick with the `BlockSkipperInterface` and also introduce a `BlockVisitor`. The visitor could do what is happening in this code passage, but we still would have to do some stuff twice (namely iterating over the `ScheduleBlockSkipper` and the `CacheLifeTimeBlockVisitor`).

Reply:
I've read a bit about the `Visitor` pattern, and I think this is quite a good match.

Actually it seems to be built for traversing entire object structures, which would also make sense in our case. It would be very nice to be able to traverse over the entire metadata of a page that way. Then we could e.g. build a `CacheLifetimeVisitor` and then every content type, not only the blocks, could have an effect on the cache lifetime. But I don't think it makes a lot of sense to build that now, especially since I am quite sure that would need a few more adjustments.

The pattern consists of 2 important elements:

1. **`Element`**: Is a part of the object structure that will be traversed. Usually has an `accept` method, which takes the visitor and calls the appropriate function.
2. **`Visitor`**: Has a `visit` method for each element that can be passed, and will do something with them.

I think the `accept` method would not be really nice in our existing design, and according to the GoF book it also is not absolutely necessary. As an alternative approach it mentions that the object structure itself iterates over all elements and call the `Visitor` themselves. This means there are less possibilities to dispatch different `Visitor` implementations, but I think that is not necessary in our case. What we will do instead is to have multiple `Visitor` implementions.

That's what I've got so far today, and I wanted to jot it down somewhere. Will continue to work on that tomorrow and hopefully come up with a design.

Reply:
The `Visitor` pattern would probably start showing of its full strength, if it would be used for the entire `StructureInterface`. If it would be used like that we could have a `CacheLifetimeVisitor`, `ContentDataVisitor` and a `ViewDataVisitor`, making the logic being split nicely and traversing the entire tree would only have to be implemented once.

But now probably is not the right time to do that, therefore I would suggest to implement something like this in the `BlockContentType` (omitting some code to keep the example brief):

```php
private function prepareData($property) {
    // ...
    for ($i = 0; $i < $blockProperty->getLength(); ++$i) {
        $blockPropertyType = $blockProperty->getProperties($i);
        foreach ($blockVisitor as $this->blockVisitors) { // strictly speaking the visitor could hold some internal state, so this could be a factory instead
            $blockPropertyType = $blockVisitor->visit($blockPropertyType);

            if (!$blockPropertyType) {
                continue 2;
            }
        }

        // read data from block
    }
}
```

In this scenario all the currently existing `BlockSkipperInterface` implementations can be replaced by a `BlockVisitor`, that returns `null` if the block should be skipped. Additionally a `CacheLifetimeBlockVisitor` could find the soonest future date something changes again, and call the `CacheLifetimeRequestEnhancer` for each block with that date.

Question is if we should introduce a `BlockVisitoryFactoryInterface` as well, to allow the `BlockVisitorInterface` implementations to have internal state as well. This would not be possible if the `BlockVisitorInterface` is registered as a service, since state would also be shared among different requests, etc.

Reply:
Code looks fine from my side. I personally would not introduce any factory as this mostly make it more complicated to create this visitors and allow to autowire them in the symfony eco system. So would just go with the `BlockVisitorInterface` and avoid the introduction of the Factory.

Reply:
In the Symfony ecosystem the `Factory` could also be autowired, so I don't think that should be a big problem.

The problem IMO is that the visitor pattern is generally understood that way, and if somebody wants to build a visitor holding some internal state in the future, we have already blocked that by this decision. But it's also ok for me to do it like this for now and live with these consequences. Would just mention it in a PHPDoc of the interface.

It's especially ok for me, since we are slightly changing the pattern already, because we are returning something from the `visit` function, which is not the case in the original pattern. There all `visit` functions would be returning nothing, and the visitor itself has another function returning the end result. That way it would maybe make more sense if we took the different approach I mentioned above (traversing the entire structure and also having a `ContentDataVisitor` and a `ViewDataVisitor`). But maybe this way we can see it as dipping our toes in the water :slightly_smiling_face: